rllib/policy/sample_batch.py

```diff
@@ -58,35 +64,58 @@ def __init__(self, *args, **kwargs):
         # Possible seq_lens (TxB or BxT) setup.
         self.time_major = kwargs.pop("_time_major", None)
         self.seq_lens = kwargs.pop("_seq_lens", None)
+        if isinstance(self.seq_lens, list):
+            self.seq_lens = np.array(self.seq_lens)
         self.dont_check_lens = kwargs.pop("_dont_check_lens", False)
         self.max_seq_len = kwargs.pop("_max_seq_len", None)
         if self.max_seq_len is None and self.seq_lens is not None and \
+                not (tf and tf.is_tensor(self.seq_lens)) and \
                 len(self.seq_lens) > 0:
             self.max_seq_len = max(self.seq_lens)
         self.zero_padded = kwargs.pop("_zero_padded", False)
+        self.is_training = kwargs.pop("_is_training", None)
+
+        # Call super constructor. This will make the actual data accessible
+        # by column name (str) via e.g. self["some-col"].
+        dict.__init__(self, *args, **kwargs)
+
+        if self.is_training is None:
+            self.is_training = self.pop("is_training", False)
+        if self.seq_lens is None:
+            self.seq_lens = self.get("seq_lens", None)
 
-        # The actual data, accessible by column name (str).
-        self.data = dict(*args, **kwargs)
+        self.accessed_keys = set()
+        self.added_keys = set()
+        self.deleted_keys = set()
+        self.intercepted_values = {}
+
+        self.get_interceptor = None
 
         lengths = []
-        for k, v in self.data.copy().items():
+        copy_ = {k: v for k, v in self.items()}
+        for k, v in copy_.items():
             assert isinstance(k, str), self
-            lengths.append(len(v))
+            len_ = len(v) if isinstance(
+                v,
+                (list, np.ndarray)) or (torch and torch.is_tensor(v)) else None
+            lengths.append(len_)
             if isinstance(v, list):
-                self.data[k] = np.array(v)
+                self[k] = np.array(v)
+
         if not lengths:
             raise ValueError("Empty sample batch")
+
         if not self.dont_check_lens:
             assert len(set(lengths)) == 1, \
                 "Data columns must be same length, but lens are " \
                 "{}".format(lengths)
-        if self.seq_lens is not None and len(self.seq_lens) > 0:
+
+        if self.seq_lens is not None and \
```

Comment:
There seems to be a lot of edge cases of handling seq_lens in this PR. Is there a better way around it?

Reply:
Good point. I'm trying to establish `seq_lens` as a property, rather than a (str) key in SampleBatch (similar to `is_training`). At the same time, this needs to be backward compatible, for when users do access it via the "seq_lens" dict key.
It may be None (or an empty list) and we do have to check for those cases.
This will get a little better, once we deprecate the "seq_lens" key and only allow access via the property.